### PR TITLE
S3 email validation

### PR DESCRIPTION
**Pull request by imogenhallett**
Created class Email to validate email address.

Created class User to check if email address is already in database and add if not.

Created class ErrorLog to write standard PHP errors to a log file.

Also setup Composer and Pokedex namespace.

Please note: test_form.php is included for testing purposes.

Please review all files

### CODE REVIEW

File: src/ErrorLog.php [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+class ErrorLog
+{
+    private const ERROR_LOG = 'errorLog.txt';
+
+    /** This logs messages to the error log
+     * @param string $error Message can be user defined or accessed from Exception object
+     */
+    public static function log(string $errorMessage) {
+        $errorLog = fopen(self::ERROR_LOG, 'a');
```
Comment (Benson-J) [inline src/ErrorLog.php]: Does not work if errorLog.txt does not already exist.

File: src/DbConnection.php [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    private $db;
+    /**
+     * DbConnection constructor creates new connection to the pokedex_hb database
+     */
+    public function __construct()
+    {
+        $this->db =  new PDO('mysql:dbname=pokedex_hb;host=127.0.0.1','root');
+    }
+
+    /** Function to get the $db property
```
Comment (Benson-J) [inline src/DbConnection.php]: No text on top line of docBlock.

File: src/ErrorLog.php [hunk 3]
```diff
@@ -0,0 +1,16 @@
+<?php
+namespace Pokedex;
+
+class ErrorLog
+{
+    private const ERROR_LOG = 'errorLog.txt';
+
+    /** This logs messages to the error log
```
Comment (Benson-J) [inline src/ErrorLog.php]: No text on top line of docBlock.

File: src/User.php [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+     */
+    public function __construct(AbstractEmail $email, \PDO $db)
+    {
+        $this->email = $email;
+
+        $stmt = $db->prepare('SELECT `id`, `email` FROM `users` WHERE `email` = :email;');
+        $stmt->execute([':email' => $this->email]);
+        $count = $stmt->rowCount();
+
+        if (!$count) {
```
Comment (Benson-J) [inline src/User.php:22]: Rather that checking if email is already in database, couldn't you just use REPLACE INTO? Which would overwrite existing emails as themselves, and create new entries for new emails.

File: src/ValueObjects/Email.php [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+     * @throws EmailException
+     */
+    public function __construct(string $email) {
+        if(!filter_var($email, FILTER_VALIDATE_EMAIL)) {
+            throw new EmailException('Invalid email address');
+        }
+        $this->email = $email;
+    }
+
+    /** Function allows the object to return the email address as a string
```
Comment (Benson-J) [inline src/ValueObjects/Email.php]: No text on top line of docBlock.

File: src/ErrorLog.php (hunk 1 above)
Comment (patrickkenyon) [reply src/ErrorLog.php]: changed to mode a+ in

Comment (patrickkenyon) [reply src/ErrorLog.php]: 66bc114

File: src/User.php (hunk 4 above)
Comment (patrickkenyon) [reply src/User.php:22]: replace into does not work in all database managers. :)

File: src/DbConnection.php (hunk 2 above)
Comment (patrickkenyon) [reply src/DbConnection.php]: changed docblocks to line below ab85232

File: src/ErrorLog.php (hunk 3 above)
Comment (patrickkenyon) [reply src/ErrorLog.php]: changed docblocks to line below ab85232

File: src/ValueObjects/Email.php (hunk 5 above)
Comment (patrickkenyon) [reply src/ValueObjects/Email.php]: added return to dockblock in Email.php 138fccb

**Review by patrickkenyon — COMMENTED** [review]
Please review and approve responses to comments

**Review by Benson-J — APPROVED** [review]

**Pull request merged**